### PR TITLE
Modernize notifier core module

### PR DESCRIPTION
**Pull request by scottinet**
# Description

Following discussions on #790, several changes have been implemented:

* split the ubiquitous `notify` method and the `NotificationObject` into more specialized versions for documents, users and server notifications
* notification methods now trigger specialized events: `notify:document`, `notify:user` and `notify:server`. Pipe plugin can update the notification content or even block it
* notification methods also trigger a global event at the very end of the process, just before sending the notification: `notify:dispatch`. Pipe plugins can update or block notifications
* notifications have been rationalized to make them more consistent (they were derived from old `RequestObject` objects). The `jwtTokenExpired` notification in particular has been changed to a `TokenExpired` specialized notification (BREAKING CHANGE)

# Boyscout

* modernized old notifier unit tests
* resolved promise rejection warnings in some unit tests
* `funnelController.processRequest`: there was an unit test on this method that was failing "silently" because of a lack of `return` statement at the end of it. This is fixed, and the bug causing this test to fail has also been fixed

# Other tasks to perform

* SDKs must be adapted to the `jwtTokenExpired` notification change
* documentation will also be updated

# Solved issue

#790

### CODE REVIEW

File: lib/api/controllers/documentController.js [hunk 1]
```diff
@@ -162,7 +162,7 @@ class DocumentController {
         modifiedRequest = newRequest;
 
         request.input.args.state = 'pending';
```
Comment (ballinette) [inline lib/api/controllers/documentController.js]: is `request.input.args.state` still necessary ?

Comment (scottinet) [inline lib/api/controllers/documentController.js]: It appears to be useless, and for a long time to boot.

Removed this assignation from all of this controller's methods.

File: lib/api/core/models/notifications/NotificationDocument.js [hunk 2]
```diff
@@ -22,78 +22,46 @@
...
+    this.scope = scope;
+    this.state = state;
+    this.result = content;
+    this.type = 'document';
   }
 
-  toJson() {
-    const object = {
-      error: null,
+  toJSON() {
```
Comment (scottinet) [inline lib/api/core/models/notifications/NotificationDocument.js]: No we don't

**Comment by scottinet** [comment]
@benoitvidis > done

Also, while documenting the newly added events, I found out that the payload was inconsistent. This is now fixed.

**Pull request merged**